plugins/videobasedtracker/UVCImageSource.cpp

```diff
@@ -245,20 +245,20 @@ namespace vbtracker {
         // upon request in the retrieveColor() method.
 
         // Copy the frame, then look for the lock
-        uvc_frame_t *copied_frame = uvc_allocate_frame(frame->data_bytes);
-        uvc_duplicate_frame(frame, copied_frame);
+        Frame_ptr copied_frame(uvc_allocate_frame(frame->data_bytes),
+                               &uvc_free_frame);
+        uvc_duplicate_frame(frame, copied_frame.get());
 
         std::lock_guard<std::mutex> lock(mutex_);
-        frames_.push(copied_frame);
+        frames_.emplace(copied_frame.release(), &uvc_free_frame);
         if (frames_.size() > 100) {
             std::cerr << "WARNING! Dropping frames from video tracker as they "
                          "are not being processed fast enough! This will "
-                         "disrupt tracking.";
-            while (!frames_.empty()) {
-                uvc_free_frame(frames_.front());
-                frames_.pop();
-            }
+                         "disrupt tracking."
+                      << std::endl;
+            frames_ = std::queue<Frame_ptr>(); //< clear the queue
```

Comment:
Thanks for the input godbyk, but I need a push_back() and pop_front() as I need a FIFO. If you implemented your own pop_front() on a std::vector, you'd have to reshuffle all the vector contents, which is additional overhead.

If I was going for speed, I'd actually use a pool of  pre-allocated frame pointers which have the linked list structure built in (i.e., like boost's intrusive containers). This would avoid the malloc/free I'm performing each frame, and give me all the container operations I need (clear is just removing the front reference).

However, this is a function which is called 100 times a second AT MOST, and performs next to no computation. I think the current code is more readable, and "premature optimisation is the root of all evil!", so lets see if any performance tests show this as a bottle neck (my bet is that the image processing is the bottleneck here).